model/roles.go

```diff
@@ -465,3 +497,247 @@ func (r *Role) calculateRoleConfigurationTemplates() {
 
 	r.Configuration.Templates = roleConfigs
 }
+
+// validateVariableSorting tests whether the parameters are properly sorted or not.
+// It reports all variables which are out of order.
+func validateVariableSorting(variables ConfigurationVariableSlice) validation.ErrorList {
+	allErrs := validation.ErrorList{}
+
+	previousName := ""
+	for _, cv := range variables {
+		if cv.Name < previousName {
+			allErrs = append(allErrs, validation.Invalid("configuration.variables",
+				previousName,
+				fmt.Sprintf("Does not sort before '%s'", cv.Name)))
+		}
+		previousName = cv.Name
+	}
+
+	return allErrs
+}
+
+// validateVariableUsage tests whether all parameters are used in a template or not.
+// It reports all variables which are not used by at least one template.
+//
+// ATTENTION: This will mis-report any variables which are used only
+// in scripts, but not in templates.
+func validateVariableUsage(roleManifest *RoleManifest) validation.ErrorList {
+	allErrs := validation.ErrorList{}
+
+	// See also 'GetVariablesForRole' (mustache.go).
+
+	unusedConfigs := MakeMapOfVariables(roleManifest)
+	if len(unusedConfigs) == 0 {
+		return allErrs
+	}
+
+	// Iterate over all roles, jobs, templates, extract the used
+	// variables. Remove each found from the set of unused
+	// configs.
+
+	for _, role := range roleManifest.Roles {
+		for _, job := range role.Jobs {
+			for _, property := range job.Properties {
+				propertyName := fmt.Sprintf("properties.%s", property.Name)
+
+				if template, ok := role.Configuration.Templates[propertyName]; ok {
+					varsInTemplate, err := parseTemplate(template)
+					if err != nil {
+						// We ignore bad templates here
+						continue
+					}
+					for _, envVar := range varsInTemplate {
+						if _, ok := unusedConfigs[envVar]; ok {
+							delete(unusedConfigs, envVar)
+						}
+						if len(unusedConfigs) == 0 {
+							// Everything got used, stop now.
+							return allErrs
+						}
+					}
+				}
+			}
+		}
+	}
+
+	// Iterate over the global templates, extract the used
+	// variables. Remove each found from the set of unused
+	// configs.
+
+	for _, template := range roleManifest.Configuration.Templates {
```

Comment:
Not sure what you mean. The global templates in the manifest are outside of the roles, and only roles have jobs associated with them. Unclear where you are going here, to me, sorry.